### PR TITLE
Added handling class data and token data

### PR DESCRIPTION
**Pull request by Art3miX**

### CODE REVIEW

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 1]
```diff
@@ -174,13 +178,13 @@ fn receive_nft(
...
-        class_uri,
-        class_data: None, // TODO: check for values we need to forward from a source chain.
+        class_id: class.id.clone(),
+        class_uri: class.uri,
+        class_data: class.data, // TODO: check for values we need to forward from a source chain.
 
         token_ids: vec![token_id.clone()],
         token_uris: token_uri.map(|uri| vec![uri]),
-        token_data: None,
+        token_data: None, // we set it to None for local NFT for now.
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:187]:
```suggestion
        token_data: None, // Like class_uri and class_data, token_data is none for local NFTs.
```

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 2]
```diff
@@ -174,13 +178,13 @@ fn receive_nft(
     )?;
 
     let ibc_message = NonFungibleTokenPacketData {
-        class_id: class_id.clone(),
-        class_uri,
-        class_data: None, // TODO: check for values we need to forward from a source chain.
+        class_id: class.id.clone(),
+        class_uri: class.uri,
+        class_data: class.data, // TODO: check for values we need to forward from a source chain.
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:183]:
```suggestion
        class_data: class.data,
```

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 3]
```diff
@@ -193,30 +197,37 @@ fn receive_nft(
...
 ) -> Result<Response, ContractError> {
     let receiver = deps.api.addr_validate(&receiver)?;
-    let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
+    let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id.clone())?;
 
     let mint = tokens
         .into_iter()
-        .map(|Token { id, uri, data: _ }| {
+        .map(|Token { id, uri, data }| {
+            // We save the metadata here.
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:224]:
```suggestion
```

Comment (0xekez) [reply contracts/cw-ics721-bridge/src/contract.rs:224]: i think this is a good call. how about a comment about _why_ you choose to save it here? one good reason is that it means we only iterate over the list once.

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 4]
```diff
@@ -293,17 +303,21 @@ fn callback_create_vouchers(
...
 ) -> Result<Response, ContractError> {
     let VoucherRedemption { class, token_ids } = redeem;
-    let nft_contract = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class.id)?;
+    let nft_contract = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class.id.clone())?;
     let receiver = deps.api.addr_validate(&receiver)?;
     Ok(Response::default().add_messages(
         token_ids
             .into_iter()
             .map(|token_id| {
+                // We remove the metadata here
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:317]: same comment about this comment. maybe:

```
// token metadata is set in mint, on voucher creation, and removed here on voucher redemption
```

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 5]
```diff
@@ -336,19 +350,17 @@ pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
     }
 }
 
-fn query_class_id_for_nft_contract(deps: Deps, contract: String) -> StdResult<Option<ClassId>> {
+fn query_class_id_for_nft_contract(deps: Deps, contract: String) -> StdResult<Option<Class>> {
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:353]:
```suggestion
fn query_class_for_nft_contract(deps: Deps, contract: String) -> StdResult<Option<Class>> {
```

File: contracts/cw-ics721-bridge/src/state.rs [hunk 6]
```diff
@@ -16,17 +16,19 @@ pub const PO: PauseOrchestrator = PauseOrchestrator::new("c", "d");
 /// contract we have instantiated for that classID.
 pub const CLASS_ID_TO_NFT_CONTRACT: Map<ClassId, Addr> = Map::new("e");
 /// Maps cw721 contracts to the classID they were instantiated for.
-pub const NFT_CONTRACT_TO_CLASS_ID: Map<Addr, ClassId> = Map::new("f");
+pub const NFT_CONTRACT_TO_CLASS: Map<Addr, Class> = Map::new("f");
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/state.rs:19]: could you say a little about why you chose to store the `Class` here instead of keeping it `ClassId` and requiring a second lookup in `CLASS_ID_TO_CLASS`. i feel a little nervous having two sources of truth for the `Class` data.

**Review by 0xekez — COMMENTED** [review]
this is really great. i can tell you thought about this! some small comments and then lg2m

File: contracts/cw-ics721-bridge/src/contract.rs (hunk 3 above)
Comment (Art3miX) [reply contracts/cw-ics721-bridge/src/contract.rs]: How about: `// We save token metadata here because it is part of the mint process`

File: contracts/cw-ics721-bridge/src/state.rs (hunk 6 above)
Comment (Art3miX) [reply contracts/cw-ics721-bridge/src/state.rs:19]: Honestly, I don't have a good reason except my logic and because of how its set up:

1. When you ask for the classId from nft contract address, chances are you also want the class uri and data, we already have a struct ready for this, so why not return the whole struct instead of just class_id? it will reduce a lot of queries.
Thats the reason why I think we should save the whole class struct, instead of just the id.
2. if you already have a class id, you might want to query just class uri and data, without `CLASS_ID_TO_CLASS` storage, you would need to query for the nft contract first, and then get the class struct?
This is why I think we need `CLASS_ID_TO_CLASS` storage.

So basically combing those 2 reasons lead to having 2 source of truth for class.

BUT looking in the code this might be wrong, because on create_voucher, we might not have the nft contract address yet:
1. so I either change `NFT_CONTRACT_TO_CLASS` back to only hold the class_id, and keep `CLASS_ID_TO_CLASS` as 1 source of truth for class uri and data (but will always require 2 queries to get class uri and data)
2. or I move both saves into the mint callback, and pass the class along to it, which is really not the place for class stuff.

Seems like option 1 is better, thoughts?

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 7]
```diff
@@ -144,28 +145,31 @@ fn receive_nft(
...
-        // We set class level metadata to None for local NFTs.
-        //
-        // Merging and usage of this PR may change that:
-        // <https://github.com/CosmWasm/cw-nfts/pull/75>
-        CLASS_ID_TO_CLASS_URI.save(deps.storage, class_id.clone(), &None)?;
-        // TODO: set metadata
-        class_id
+    let class = NFT_CONTRACT_TO_CLASS.may_load(deps.storage, info.sender.clone())?;
+
+    let class = match class {
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:150]:
```suggestion
    let class = match NFT_CONTRACT_TO_CLASS.may_load(deps.storage, info.sender.clone())? {
```

`match class` is probably less descriptive than just matching on what were doing. no need for indirection here imo

File: contracts/cw-ics721-bridge/src/ibc_packet_receive.rs [hunk 8]
```diff
@@ -160,7 +160,13 @@ pub(crate) fn receive_ibc_packet(
         )
         .into_submessage(env.contract.address, receiver)?;
 
-    Ok(IbcReceiveResponse::default()
+    let mut response = IbcReceiveResponse::default();
+
+    if data.memo.is_some() {
+        response = response.add_attribute("ics721_memo", data.memo.unwrap());
+    }
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/ibc_packet_receive.rs:167]:
```
let response = if let Some(memo) = data.memo {
   Response::default().add_attribute("memo", memo)
} else {
  Response::default()
}
```
nice to avoid making variables mutable where we don't have to. mutability is viral so where possible i try to avoid it

File: contracts/cw-ics721-bridge/src/state.rs (hunk 6 above)
Comment (0xekez) [reply contracts/cw-ics721-bridge/src/state.rs:19]: i think (1) is a good idea. nice catch

**Review by 0xekez — COMMENTED** [review]
looking good! would you mind adding some tests for these?

i think the big things that need to be covered are:

1. does the validation logic for `NonFungibleTokenPacketData` work  (unit tests + 1 integration test).
2. does the memo get added to response attributes (unit test or heavily mocked integration test b/c multi-test can't do IBC stuff)
3. does class data get forwarded? this should be testable using the adversarial tests framework by crafting a packet on the adversary chain with data and then sending it across the other two and making sure that the data is queryable.

File: contracts/cw-ics721-bridge/src/ibc_packet_receive.rs (hunk 8 above)
Comment (Art3miX) [reply contracts/cw-ics721-bridge/src/ibc_packet_receive.rs:167]: nice, need to get used to the let if let thing

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 9]
```diff
@@ -345,10 +360,8 @@ fn query_nft_contract_for_class_id(deps: Deps, class_id: String) -> StdResult<Op
     CLASS_ID_TO_NFT_CONTRACT.may_load(deps.storage, ClassId::new(class_id))
 }
 
-fn query_metadata(deps: Deps, class_id: String) -> StdResult<Option<String>> {
-    Ok(CLASS_ID_TO_CLASS_URI
-        .may_load(deps.storage, ClassId::new(class_id))?
-        .flatten())
+fn query_metadata(deps: Deps, class_id: String) -> StdResult<Option<Class>> {
+    CLASS_ID_TO_CLASS.may_load(deps.storage, ClassId::new(class_id))
 }
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:365]: the return type annotation for `QueryMsg::Metadata {}` needs to be updated to return `Option<Class>` as well

File: contracts/cw-ics721-bridge/src/msg.rs [hunk 10]
```diff
@@ -90,6 +90,8 @@ pub struct IbcOutgoingMsg {
     pub channel_id: String,
     /// Timeout for the IBC message.
     pub timeout: IbcTimeout,
+    /// Memo to add custom string to the msg
+    pub memo: Option<String>,
 }
 
 #[cw_serde]
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/msg.rs:97]: we need to add a query to this to make the new token metadata accessible. i'd suggest renaming `Metadata { class_id }` to `ClassMetadata { class_id }` and adding an additional `TokenMetadata { class_id, token_id }` query.

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 11]
```diff
@@ -174,16 +178,17 @@ fn receive_nft(
...
-        class_uri,
-        class_data: None, // TODO: check for values we need to forward from a source chain.
+        class_id: class.id.clone(),
+        class_uri: class.uri,
+        class_data: class.data,
 
         token_ids: vec![token_id.clone()],
         token_uris: token_uri.map(|uri| vec![uri]),
-        token_data: None,
+        token_data: None, // Like class_uri and class_data, token_data is none for local NFTs.
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:187]: need for forward token metadata here!

```rust
    let token_metadata = CLASS_TOKEN_ID_TO_TOKEN_METADATA
        .may_load(deps.storage, (class.id.clone(), token_id.clone()))?
        .flatten();
```

File: contracts/cw-ics721-bridge/src/contract.rs [hunk 12]
```diff
@@ -293,17 +304,21 @@ fn callback_create_vouchers(
...
-    let nft_contract = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class.id)?;
+    let nft_contract = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class.id.clone())?;
     let receiver = deps.api.addr_validate(&receiver)?;
     Ok(Response::default().add_messages(
         token_ids
             .into_iter()
             .map(|token_id| {
+                // token metadata is set in mint, on voucher creation, and removed here on voucher redemption
+                CLASS_TOKEN_ID_TO_TOKEN_METADATA
+                    .remove(deps.storage, (class.id.clone(), token_id.clone()));
```
Comment (0xekez) [inline contracts/cw-ics721-bridge/src/contract.rs:320]: this would clear the token metadata on the source chain if a sink chain returned a debt voucher. it seems to me that we actually want to clear this on the sink chain. otherwise, sending a debt-voucher away and returning it would cause its metadata to be lost.

it seems to me that the appropriate place to do this is in `ibc_packet_ack` when we burn the token if it is returning to its source.

**Comment by 0xekez** [comment]
@Art3miX, here are fixes for the comments i added today: https://github.com/Art3miX/ics721/pull/2

merge at your leisure.

**Comment by 0xekez** [comment]
merging this. i will write more tests on top. thanks @Art3miX! you have significantly increased the speed at which we'll be able to ship this.
